synapse/storage/databases/main/event_federation.py

```diff
@@ -922,6 +962,26 @@ def _get_backfill_events(self, txn, room_id, event_list, limit):
                 if row[1] not in event_results:
                     queue.put((-row[0], row[1]))
 
+            # From each event, go forwards through successors to find chains of
+            # historical backfilled floating outliers
+            successorQueue = PriorityQueue()
+            successorQueue.put(event_id)
+            while not successorQueue.empty() and len(event_results) < limit:
+                try:
+                    event_id = successorQueue.get_nowait()
+                except Empty:
+                    break
+
+                successor_event_ids = self.get_successor_events_txn(txn, [event_id])
+                for successor_event_id in successor_event_ids:
+                    # Skip any branches we have already gone down
+                    if successor_event_id in event_results:
+                        continue
+
+                    event_results.add(successor_event_id)
+
+                    successorQueue.put(successor_event_id)
+
```

Comment:
Summary from [another thread](https://github.com/matrix-org/synapse/pull/9247#discussion_r622059435) about making sure new historical messages are still visible after a federated server already has scrollback history of everything:

> The other issue that's likely to crop up here is what happens when a server joins, gets _all_ the history, and _then_ the original HS adds some historic events. The remote server won't ever call `/backfill` again as it thinks it has all the events in the room (since it won't have anything in its `event_backward_extremities` table, which is what governs when and with what to call `/backfill` with).
> 
> I think what we'd want to do is to force a gap at where we'd want to insert history, that way whenever another server reaches that point it'll always try and backfill there. When the `_get_backfill` function is called on the original server it knows that if it reaches that gap to go and pull in another chunk of historical events, iyswim? Downside to this is that only the original server would know to how to handle `/backfill` there.
> 
> We could also add an event there named "insertion point" which contains the server and an opaque ID for the next "chunk" of historical events to be pulled. I.e. when a server reaches that point, and hasn't yet received the chunk, it can go and request it from the server (which could then dynamically load it in). Or something
>
> *-- @erikjohnston, https://github.com/matrix-org/synapse/pull/9247#discussion_r622059435*

---

> Can we use the push part of the federation whenever new historical messages come in? (hs1 calls `PUT hs2/_matrix/federation/v1/send/{txnId}`)
> 
> Seems like if `hs2` is already joined to the room, we can just fill them in when `hs1` gets historical messages added as well. And then this avoids any need for "gaps" or "insertion points"
> 
> *-- @MadLittleMods, https://github.com/matrix-org/synapse/pull/9247#discussion_r622796142*


---

> We tend towards not having all history on all servers, so I'm not hugely keen to actively push historical stuff that half the servers probably don't care about. It's an option though.
>
> *-- @erikjohnston, https://github.com/matrix-org/synapse/pull/9247#discussion_r623260889*

---

> Using push federation feels like the least complex solution.
> 
> The gap and insertion point also feel weird to me since they rely on the original homeserver as the central source of truth which goes against the whole decentralized ethos.
> 
> *-- @MadLittleMods, https://github.com/matrix-org/synapse/pull/9247#discussion_r624432973*

Reply:
Added a Complement test for this and it is an problem :+1:

@erikjohnston Thanks for discussion so far on this! 😀

Could it be some push federation endpoint that marks the event we inserted our historical messages next to as an `event_backward_extremities` again? That way, we don't have to push over all of the history, but the federated server knows there is something to fetch when someone scrolls back.

This sorta sounds like the "gap" but I am not sure how to insert the "gap" on the federated server if we already have a problem with federated server knowing there are historical messages to fetch.

Reply:
The main thing I'm trying to wrap my head around is if we'll need to *also* do the "gap" even if in a push style model. In the push model we need to deal with servers going offline/coming back, new servers joining servers leaving and rejoining (and so having a bunch of old history), etc.

> The gap and insertion point also feel weird to me since they rely on the original homeserver as the central source of truth which goes against the whole decentralized ethos.

Fair, though its worth noting that this is specifically for the case where we *do* have a central server inserting new history dynamically.

Reply:
> servers going offline/coming back

🤕 oof, that is tricky.

Could the gap event be sent as a "live" event to mark a previous event as `event_backward_extremities` again?

That way the federated server can pick up the gap when it comes online again and learns it needs to get some historical when someone scrolls back again.

----

> Fair, though its worth noting that this is specifically for the case where we do have a central server inserting new history dynamically.

For example, Gitter is just sending the historical messages, but once it's federated to other servers, they can also federate it. The insertion point keeps everything pointing at a bit rotted dead server in the future.

Reply:
A big explainer comment with potential comments added to the MSC (https://github.com/matrix-org/matrix-doc/pull/2716/files#r630754564) which dives deeper into this topic and summarizes the [sync call](https://docs.google.com/document/d/1JJDuPfcPNX75fprdTWlxlaKjWOdbdJylbpZ03hzo638/edit) @erikjohnston and I had around this topic.